operator/iamkeycontroller/create.go

```diff
@@ -54,45 +64,110 @@ func (p *IAMKeyPipeline) Create(ctx context.Context, mg resource.Managed) (manag
 
 // createIAMKey creates a new IAMKey in the project associated with the API Key and Secret.
 func (p *IAMKeyPipeline) createIAMKey(ctx *pipelineContext) error {
+
 	iamKey := ctx.iamKey
 	log := controllerruntime.LoggerFrom(ctx)
+	log.Info("starting creation")
+
+	policyAllow := exooapi.IamServicePolicyRuleActionAllow
+	policyDeny := exooapi.IamServicePolicyRuleActionDeny
+	policyRules := exooapi.IamServicePolicyTypeRules
+
+	log.Info("IAM Role doesnt exists, creating", "keyName", ctx.iamKey.Spec.ForProvider.KeyName)
+	autogeneratedAppcatRole := &exooapi.IamRole{
+		Name:        &iamKey.Spec.ForProvider.KeyName,
+		Description: pointer.String("IAM Role for SOS+IAM creation, it was autogenerated by provider-exoscale"),
+		Permissions: &[]exooapi.IamRolePermissions{
+			exooapi.IamRolePermissionsBypassGovernanceRetention,
+		},
+		Editable: pointer.Bool(true),
+		Policy: &exooapi.IamPolicy{
+			DefaultServiceStrategy: exooapi.IamPolicyDefaultServiceStrategyDeny,
+			Services: exooapi.IamPolicy_Services{
+				AdditionalProperties: map[string]exooapi.IamServicePolicy{
+					"sos": {
+						Type: &policyRules,
+						Rules: &[]exooapi.IamServicePolicyRule{
+							{
+								Action: &policyDeny,
+								// bucket name in this rule must be in single quotes, otherwise it doesn't work and user has access to everything
+								Expression: pointer.String("resources.bucket != " + "'" + strings.Split(iamKey.Spec.ForProvider.KeyName, ".")[1] + "'"),
+							},
+							{
+								Action:     &policyAllow,
+								Expression: pointer.String("true"),
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	// send request
+	resp, err := ExecuteRequest(ctx, "POST", ctx.iamKey.Spec.ForProvider.Zone, "/v2/iam-role", p.apiKey, p.apiSecret, autogeneratedAppcatRole)
+	if err != nil {
+		return err
+	}
+
+	iamRole := exooapi.Operation{}
+
+	// unmarshall response to iamRoles
+	err = json.NewDecoder(resp.Body).Decode(&iamRole)
+	if err != nil {
+		log.Error(err, "Cannot unmarshall response to iamRolesResponse")
+		return err
+	}
+
+	iamRoleID := *iamRole.Reference.Id
+	log.Info("IAM Role created", "iamRoleID", iamRoleID)
+
+	log.Info("IAM Key doesnt exists, creating", "keyName", ctx.iamKey.Spec.ForProvider.KeyName)
+	newIamKey := exooapi.IamApiKey{
+		RoleId: &iamRoleID,
+		Name:   &ctx.iamKey.Spec.ForProvider.KeyName,
+	}
+
+	b1, _ := json.MarshalIndent(newIamKey, "\t", "\t")
```

Comment:
Please check the code before sending for a review. There's a lost of debug code everywhere